### DISCUSSION TITLE
Can Custom Layouts handle 'resize pane' operations?

### DISCUSSION THREAD

**Discussion by thejettdurham** [ianyh/Amethyst#1480] [Q&A] [answered]
Hello 👋 ! I started experimenting with Custom Layouts this morning wand was able to eventually get something working close to how I wanted. But I was left with one burning question: **is it possible to handle the 'shrink/expand main pane' commands in a custom layout**?

I would expect to be able to handle these in `updateWithChange()`, but when I send these commands to a custom layout, I see that the change type is always `layout_change` with no extra arguments. I see that I could implement a similar functionality with custom commands, but I just want to make sure I'm not missing something obvious before going down that road. My muscle memory would really appreciate being able to handle _resize pane_ operations directly in a custom layout 😁 

Lastly I just wanted to say: I've been using Amethyst for about a year and it's been a very valuable addition to my productivity suite, so thanks to all maintainers for their hard work in building and maintaining this excellent software!

**Comment by ianyh** [comment, marked as answer]
The intention is to use the custom commands for things like that. I think I might add something to redirect standard commands when extending a layout, but that wouldn't be an immediate change.

**Reply by rwjack** [reply]
Hi, could you please explain how to name the custom commands in a custom layout?

I'm cooking something up with cgpt but the resize binds don't work (which now I know why), so I've removed those and added the same binds to the Custom layout command 1..4 binds, yet I'm still not able to resize anything. Here's the code:

```js
function layout() {
  return {
    name: "Centered Widescreen",
    initialState: {
      mainPaneCount: 1,
      mainPaneRatio: 0.5,
    },
    getFrameAssignments: (windows, screenFrame, state) => {
      const totalWindows = windows.length;
      const mainPaneCount = Math.min(state.mainPaneCount, totalWindows);
      const mainPaneRatio = state.mainPaneRatio;

      if (totalWindows === 0) return {};

      const mainPaneWidth = screenFrame.width * mainPaneRatio;
      const mainPaneHeight = screenFrame.height;
      const secondaryPaneWidth = screenFrame.width - mainPaneWidth;
      const secondaryPaneHeight = screenFrame.height / (totalWindows - mainPaneCount);

      let frames = {};

      // Assign frames to main pane windows
      for (let i = 0; i < mainPaneCount; i++) {
        frames[windows[i].id] = {
          x: screenFrame.x + mainPaneWidth * i,
          y: screenFrame.y,
          width: mainPaneWidth / mainPaneCount,
          height: mainPaneHeight,
        };
      }

      // Assign frames to secondary pane windows
      for (let i = mainPaneCount; i < totalWindows; i++) {
        frames[windows[i].id] = {
          x: screenFrame.x + mainPaneWidth,
          y: screenFrame.y + secondaryPaneHeight * (i - mainPaneCount),
          width: secondaryPaneWidth,
          height: secondaryPaneHeight,
        };
      }

      return frames;
    },
    commands: {
      customLayoutCommand1: {
        description: "Shrink main pane",
        updateState: (state) => ({
          ...state,
          mainPaneRatio: Math.max(0.1, state.mainPaneRatio - 0.05),
        }),
      },
      customLayoutCommand2: {
        description: "Expand main pane",
        updateState: (state) => ({
          ...state,
          mainPaneRatio: Math.min(0.9, state.mainPaneRatio + 0.05),
        }),
      },
      customLayoutCommand3: {
        description: "Increase main pane count",
        updateState: (state) => ({
          ...state,
          mainPaneCount: Math.min(state.mainPaneCount + 1, 3),
        }),
      },
      customLayoutCommand4: {
        description: "Decrease main pane count",
        updateState: (state) => ({
          ...state,
          mainPaneCount: Math.max(state.mainPaneCount - 1, 1),
        }),
      },
    },
  };
}
```

**Reply by rwjack** [reply]
@ianyh looking at https://github.com/ianyh/Amethyst/pull/1735, I shouldn't even have to use custom commands and could use the default binds?

**Comment by kerryj89** [comment]
Would this allow resizing windows using the mouse when/if implemented or is that another issue that I should make a request for?  I use the keyboard shortcuts for everything else, but as a web designer and developer I am often needing to test arbitrary window sizes.
